enigma-js/src/enigma-utils.js

```diff
@@ -202,12 +204,14 @@ function appendMessages(hexStr, inputsArray) {
  *
  * @param {string} hexStr - Buffer being appended to
  * @param {Array} inputsArray - Array of array of inputs
+ * @param {boolean} principal - Principal hashing
  * @return {string} - Final appended hex string
  */
-function appendArrayMessages(hexStr, inputsArray) {
+function appendArrayMessages(hexStr, inputsArray, principal=false) {
+  const principalPrefix = principal ? '01' : '';
   for (let array of inputsArray) {
-    hexStr += JSBI.BigInt(array.length).toString(16).padStart(16, '0');
-    hexStr = appendMessages(hexStr, array);
+    hexStr += principalPrefix + JSBI.BigInt(array[0].length * (array[1]+9)).toString(16).padStart(16, '0');
```

Comment:
Please see the comment above about the harcoded solution.

Reply:
Please see my answer above about the hardcoded solution :)